# AC 221: Critical Thinking in Data Science

## Problem Set 1, Part 2
**February 27, 2020**<br/>
**Harvard University**<br/>
**Spring 2020**<br/>

<hr style="height:2pt">

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import random
from azureml.opendatasets import UsPopulationZip

ModuleNotFoundError: No module named 'azureml'

## Problem 1

### Data Sources
#### US Census FactFinder
The distribution of age was calculating using US 2010 Decennial Census.

Annual Estimates of the Resident Population by Single Year of Age and Sex for the United States: April 1, 2010 to July 1, 2018
Source: U.S. Census Bureau, Population Division
Release Date: June 2019 
Link: https://factfinder.census.gov/faces/tableservices/jsf/pages/productview.xhtml?src=bkmk

#### Microsoft Azure - Open Datasets
The distributions for zipcode, sex, and gender were calculated using Microsoft Azure's *US Population by Zipcode* Open Dataset, which in turn was sourced from the 2010 US Decennial Census. Because the census collects data only on sex and not gender, we used sex as a proxy for gender.

Link: https://azure.microsoft.com/en-us/services/open-datasets/catalog/us-decennial-census-zip/

### Age

In [ ]:
age_df = pd.read_csv("data/census_age_0-100.csv")
age_2010 = age_df[['Age', '2010']][1:]

In [ ]:
xticks = list(range(0, 100, 5))
xtick_labels = [str(x) for x in xticks]
xticks.append(100)
xtick_labels.append('100+')

plt.figure(figsize=(10,10))
plt.title("Figure 1.1: Histogram of Age in United States, 2010")
plt.bar(age_2010['Age'], age_2010['2010'], width = 1.0)
plt.xlabel("Age")
plt.ylabel("Frequency")
plt.xticks(xticks, xtick_labels)
plt.show()

As seen in **Figure 1.1**, age is not uniformly distributed in the United States. Instead, it has a bimodal, right skewed distribution. 

### Zip Code

In [ ]:
population_df = UsPopulationZip().to_pandas_dataframe()

In [ ]:
# Remove rows with None's, which are aggregations of the more specific values.
pop = population_df.query('~(sex.isnull()) & (~race.isnull()) & ~(minAge.isna() | maxAge.isna())', engine='python')

In [ ]:
# Graph
ax = pop.groupby('zipCode')['population'].sum().plot.hist(bins=50, logy=True)
ax.set_title('Figure 1.2: Histogram of Log Pop. Counts within Zipcodes, 2010')
ax.set_xlabel('Pop. Size')

**Figure 1.2** shows that the population size within zipcodes is not uniformly distributed.

### Gender (By Proxy of Sex)

In [ ]:
sex_ratios = pop.groupby(['zipCode']).apply((lambda g: g.query('sex == "Female"')['population'].sum()
                                       / g.query('sex == "Male"')['population'].sum()))
np.random.seed(1)
ax = sex_ratios.replace([np.inf, -np.inf], np.nan).dropna().sample(1000).plot.hist(bins=30)
ax.set_xlabel("Female to Male Population Ratio")
ax.set_title("Figure 1.3: Histogram of Female to Male Population Ratio by Zipcode, 2010")

Interpretation.

### Race

In [ ]:
cambridge = pop.query('zipCode == "02138"')
ax = pd.DataFrame({
    'US Total': pop.groupby('race')['population'].sum() / 
                pop['population'].sum(),
    'Cambridge (02138)': cambridge.groupby('race')['population'].sum() / 
                         cambridge['population'].sum(),
}).plot.bar(logy=True)
ax.set_title('Figure 1.4: US vs. Cambridge Race Proportions')
ax.set_ylabel('Proportion of Total Population (log scale)')

Interpretation.

## General edX Data Processing

In [2]:
quasi = pd.read_csv('data/final_quasi-identifiers.csv')
quasi_list = list(quasi['variable'].values)

# Drop user_id from quasi_list
quasi_list.remove('user_id')

In [3]:
df = pd.read_csv('APCOMP221Pset1.csv').drop(['username', 'ip', 
                                             'un_major_region', 'un_economic_group', 
                                             'un_developing_nation', 'un_special_region',
                                             'latitude', 'longitude', 'profile_country'], axis=1)

/home/pau/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3051: DtypeWarning: Columns (49,50,64) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
# Read small dataset for writing functions
df = pd.read_csv('APCOMP221Pset1.csv', nrows = 10000).drop(['username', 'ip', 
                                             'un_major_region', 'un_economic_group', 
                                             'un_developing_nation', 'un_special_region',
                                             'latitude', 'longitude', 'profile_country'], axis=1)

### Replacing NaNs

In [4]:
na_replacements = {
    'course_id': 'NA',
    'cc_by_ip': 'NA',
    'countryLabel': 'NA',
    'continent': 'NA',
    'city': 'NA',
    'region': 'NA',
    'subdivision': 'NA',
    'postalCode': 'NA',
    'LoE': 'NA',
    'YoB': -1,
    'gender': 'NA',
    'nforum_posts': 0,
    'nforum_votes': 0,
    'nforum_endorsed': 0,
    'nforum_threads': 0,
    'nforum_comments': 0,
    'nforum_pinned': 0,
    'nforum_events': 0,
    'mode': 'NA',
    'email_domain': 'NA'
}

df = df.fillna(na_replacements)

df.head()

course_id  user_id  registered  viewed explored  certified  \
0  HarvardX/PH525.1x/1T2018  1488411        True   False      NaN      False   
1  HarvardX/PH525.1x/1T2018  7013084        True    True     True      False   
2  HarvardX/PH525.1x/1T2018  4083257        True    True    False      False   
3  HarvardX/PH525.1x/1T2018  4605571        True    True    False      False   
4  HarvardX/PH525.1x/1T2018  1499820        True    True     True      False   

   completed cc_by_ip    countryLabel      continent  ... roles_isInstructor  \
0      False       GB  United Kingdom         Europe  ...                NaN   
1      False       US   United States       Americas  ...                NaN   
2      False       NA              NA             NA  ...                NaN   
3      False       US   United States  North America  ...                NaN   
4      False       HU         Hungary         Europe  ...                NaN   

  roles_isStaff roles_isCCX roles_isFinance roles_isLibrary  roles_isSales  \
0           NaN         NaN             NaN             NaN            NaN   
1           NaN         NaN             NaN             NaN            NaN   
2           NaN         NaN             NaN             NaN            NaN   
3           NaN         NaN             NaN             NaN            NaN   
4           NaN         NaN             NaN             NaN            NaN   

  forumRoles_isAdmin  forumRoles_isCommunityTA  forumRoles_isModerator  \
0                NaN                       NaN                     NaN   
1                NaN                       NaN                     NaN   
2                NaN                       NaN                     NaN   
3                NaN                       NaN                     NaN   
4                NaN                       NaN                     NaN   

  forumRoles_isStudent  
0                    1  
1                    1  
2                    1  
3                    1  
4                    1  

[5 rows x 83 columns]

## Problem 3

### Useful Functions

In [5]:
def suppress_below_k(k):
    groups = df.groupby(quasi_list).size()
    groups_below_threshold = groups[groups < k]
    return groups[groups < k]

def suppressed_n(groups):
    return groups_below_threshold.sum()

def percent_remaining(groups):
    return (1 - groups_below_threshold.sum() / df.shape[0]) * 100

def suppress_to_k_anonymize(df, quasi_ids, k):
    return df.groupby(quasi_ids).filter(lambda g: g.shape[0] >= k)

def status_count(df, status):
    return df.query(status + ' == True').shape[0]

def status_proportion(df, status):
    return status_count(df, status) / df.shape[0] * 100

def print_status_stats(df):
    print(f'Total students: {df.shape[0]}\n'
          f'Students explored: {status_count(df, "explored")}'
          f' - {status_proportion(df, "explored"):.2f}%\n'
          f'Students completed: {status_count(df, "completed")}'
          f' - {status_proportion(df, "completed"):.2f}%')

### Preliminary Check of User ID

In [23]:
### Checking if any user takes the same course more than once
sizes = df.groupby('user_id')['course_id'].size()
unique_sizes = df.groupby('user_id')['course_id'].nunique()
compare = (sizes == unique_sizes)
same_class = compare[compare == True]
same_class.value_counts()

True    936434
Name: course_id, dtype: int64

It does not appear that any user is taking the same class more than once, which implies that all $n$ records for the same quasi-identifier are from different users. 

### Raw Dataset Stats

In [8]:
print_status_stats(df)

Total students: 1141735
Students explored: 122258 - 10.71%
Students completed: 28142 - 2.46%


### 3-Anonymous Results

In [18]:
groups_below_threshold = suppress_below_k(3)

print("k = 3")
print("N suppresed: {}".format(suppressed_n(groups_below_threshold)))
print("Percent remaining: {}".format(percent_remaining(groups_below_threshold)))

k = 3
N suppresed: 988191
Percent remaining: 13.448304554033996


In [ ]:
k3_df = suppress_to_k_anonymize(df, quasi_list, 3)
print_status_stats(k3_df)

When making the quasi-identifier dataset k-anonymous using suppression for $k=3$, 988191 records are suppressed. This leaves **13.45%** of the 1141735 records in the original dataset. 

The completion rate is X. The exploration rate is Y. 

### 4-Anonymous Results

In [20]:
# k = 4
groups_below_threshold = suppress_below_k(4)

print("k = 4")
print("N suppresed: {}".format(suppressed_n(groups_below_threshold)))
print("Percent remaining: {}".format(percent_remaining(groups_below_threshold)))

k = 4
N suppresed: 1010127
Percent remaining: 11.527018090887985


In [ ]:
k4_df = suppress_to_k_anonymize(df, quasi_list, 4)
print_status_stats(k3_df)

When making the quasi-identifier dataset k-anonymous using suppression for $k=4$, 1010127 records are suppressed. This leaves **11.53%** of the 1141735 records in the original dataset. 

The completion rate is X. The exploration rate is Y.

In [ ]:
# k = 5
groups_below_threshold = suppress_below_k(5)

print("k = 5")
print("N suppresed: {}".format(suppressed_n(groups_below_threshold)))
print("Percent remaining: {}".format(percent_remaining(groups_below_threshold)))

In [ ]:
k4_df = suppress_to_k_anonymize(df, quasi_list, 5)
print_status_stats(k3_df)

When making the quasi-identifier dataset k-anonymous using suppression for $k=4$, 1023587 records are suppressed. This leaves **10.35%** of the 1141735 records in the original dataset. 

The completion rate is X. The exploration rate is Y.

## Question 4

### Useful Functions

In [6]:
def n_synthetic_required_for(k):
    groups = df.groupby(quasi_list).size()
    groups_below_threshold = groups[groups < k]
    return (k - groups_below_threshold).sum()

def suppressed_records(df, quasi_ids, k):
    return df.groupby(quasi_ids).filter(lambda g: g.shape[0] < k)

def free_ids(df, n):
    # Create list of unused IDs for synthetic records
    user_ids = df['user_id'].unique()
    user_ids.sort()

    range_ids = np.arange(1,10000001, 1)
    mask = np.isin(range_ids, user_ids, invert = True)
    free_ids = range_ids[mask]
    return random.choices(free_ids, k = n)

def marginal_variable_values(df, variable, n):
    # Create list of random values for desired variable using distribution from full dataset
    counts = df[variable].value_counts()
    return random.choices(counts.index.values, weights=counts.values, k=n)

def variables_for_synthetic(variables):
    # Create list of all variables in synthetic dataset: only care about quasi-identifiers, user_id, completed, and explored
    suppressed_vars = quasi_list.copy()
    for var in variables:
        suppressed_vars.append(var)
    return suppressed_vars

def suppressed_multi_index(df, quasi_identifiers, k):
    # Get multi-index
    groups = df.groupby(quasi_identifiers).size()
    return groups[groups < k]

def split_df(df, variables_for_synthetic, quasi_identifiers, k):
    # Split dataframe into suppressed and non-suppressed records
    suppressed_df = suppressed_records(df[variables_for_synthetic], quasi_list, k)
    nonsuppressed_df = suppress_to_k_anonymize(df[variables_for_synthetic], quasi_list, k)
    return suppressed_df, nonsuppressed_df

def synthetic_df(df, non_quasi_variables, quasi_identifiers, k):
    # Setup
    synth_n = n_synthetic_required_for(k)
    all_synth_vars = variables_for_synthetic(non_quasi_variables)
    sup_df, nonsup_df = split_df(df, all_synth_vars, quasi_identifiers, k)
    
    # Random lists
    random_user_ids = free_ids(df, synth_n)
    random_completed = marginal_variable_values(df, 'completed', synth_n)
    random_explored = marginal_variable_values(df, 'explored', synth_n)
    
    # Indices setup for synthetic loop
    groups_below_threshold = suppressed_multi_index(df, quasi_identifiers, k)
    sup_df.set_index(quasi_list, inplace = True)
    index_tuples = tuple(zip(groups_below_threshold, groups_below_threshold.index))

    # Loop for creating synthetic records
    for t in index_tuples:
        t_size = t[0]

        # Add synthetic records
        for s in range(0, k-t_size):
            # Create synthetic variables
            user_id = random_user_ids.pop(0)
            completed = random_completed.pop(0)
            explored = random_explored.pop(0)

            # Create row
            row = list(t)
            row.append(user_id)
            row.append(completed)
            row.append(explored)
            row_s = pd.Series(row)
            sup_df = sup_df.append(row_s, ignore_index = True)
    
    # Concat suppressed (with added synthetic rows) and non-suppressed dataframes
    synth_df = nonsup_df.append(sup_df)
    
    # Randomize row order 
    return synth_df.sample(frac=1).reset_index(drop=True)

### Creating Synthetic Rows for Database
For every record whose quasi-identifiers appeared in  $n$ < $k$ records, we generated $k - n$ synthetic records.

As we are only concerned with comparing `completed` and `explored` rates, we only included these variables in addition to the quasi-identifiers and `user_id`. 

A fake `user_id` was created by randomly selecting a n unused number in the range [1, 1000000]. This was the range of numbers for real ids. This means all synthetic records represent users who did not take more than one class. Ideally, we would sample user_ids in a way that distribution of number of classes taken by a user. 

The values for `explored` and `completed` were drawn using the marginal distribution. Thus, they were each independently drawn according to the distribution of the corresponding column in the original dataset. Because our two variables are binary categories, the distributions were Bernoulli with proportions of success easy to calculate. The marginal distribution was chosen because it gives a good mean vs correlation bias trade-off while not being as computationally expensive as using the marginal mean distribution.

### 3-Anonymous Results

In [7]:
synth_k3_df = synthetic_df(df, ['user_id', 'completed', 'explored'], quasi_list, 3)
print_status_stats(synth_k3_df)

(9248, 8)
(9249, 8)
(9250, 8)
(9251, 8)
(9252, 8)
(9253, 8)
(9254, 8)
(9255, 8)
(9256, 8)
(9257, 8)
(9258, 8)
(9259, 8)
(9260, 8)
(9261, 8)
(9262, 8)
(9263, 8)
(9264, 8)
(9265, 8)
(9266, 8)
(9267, 8)
(9268, 8)
(9269, 8)
(9270, 8)
(9271, 8)
(9272, 8)
(9273, 8)
(9274, 8)
(9275, 8)
(9276, 8)
(9277, 8)
(9278, 8)
(9279, 8)
(9280, 8)
(9281, 8)
(9282, 8)
(9283, 8)
(9284, 8)
(9285, 8)
(9286, 8)
(9287, 8)
(9288, 8)
(9289, 8)
(9290, 8)
(9291, 8)
(9292, 8)
(9293, 8)
(9294, 8)
(9295, 8)
(9296, 8)
(9297, 8)
(9298, 8)
(9299, 8)
(9300, 8)
(9301, 8)
(9302, 8)
(9303, 8)
(9304, 8)
(9305, 8)
(9306, 8)
(9307, 8)
(9308, 8)
(9309, 8)
(9310, 8)
(9311, 8)
(9312, 8)
(9313, 8)
(9314, 8)
(9315, 8)
(9316, 8)
(9317, 8)
(9318, 8)
(9319, 8)
(9320, 8)
(9321, 8)
(9322, 8)
(9323, 8)
(9324, 8)
(9325, 8)
(9326, 8)
(9327, 8)
(9328, 8)
(9329, 8)
(9330, 8)
(9331, 8)
(9332, 8)
(9333, 8)
(9334, 8)
(9335, 8)
(9336, 8)
(9337, 8)
(9338, 8)
(9339, 8)
(9340, 8)
(9341, 8)
(9342, 8)
(9343, 8)
(9344, 8)
(9345, 8)
(9346, 8)
(9347, 8)


(10068, 8)
(10069, 8)
(10070, 8)
(10071, 8)
(10072, 8)
(10073, 8)
(10074, 8)
(10075, 8)
(10076, 8)
(10077, 8)
(10078, 8)
(10079, 8)
(10080, 8)
(10081, 8)
(10082, 8)
(10083, 8)
(10084, 8)
(10085, 8)
(10086, 8)
(10087, 8)
(10088, 8)
(10089, 8)
(10090, 8)
(10091, 8)
(10092, 8)
(10093, 8)
(10094, 8)
(10095, 8)
(10096, 8)
(10097, 8)
(10098, 8)
(10099, 8)
(10100, 8)
(10101, 8)
(10102, 8)
(10103, 8)
(10104, 8)
(10105, 8)
(10106, 8)
(10107, 8)
(10108, 8)
(10109, 8)
(10110, 8)
(10111, 8)
(10112, 8)
(10113, 8)
(10114, 8)
(10115, 8)
(10116, 8)
(10117, 8)
(10118, 8)
(10119, 8)
(10120, 8)
(10121, 8)
(10122, 8)
(10123, 8)
(10124, 8)
(10125, 8)
(10126, 8)
(10127, 8)
(10128, 8)
(10129, 8)
(10130, 8)
(10131, 8)
(10132, 8)
(10133, 8)
(10134, 8)
(10135, 8)
(10136, 8)
(10137, 8)
(10138, 8)
(10139, 8)
(10140, 8)
(10141, 8)
(10142, 8)
(10143, 8)
(10144, 8)
(10145, 8)
(10146, 8)
(10147, 8)
(10148, 8)
(10149, 8)
(10150, 8)
(10151, 8)
(10152, 8)
(10153, 8)
(10154, 8)
(10155, 8)
(10156, 8)
(10157, 8)
(10158, 8)

(10815, 8)
(10816, 8)
(10817, 8)
(10818, 8)
(10819, 8)
(10820, 8)
(10821, 8)
(10822, 8)
(10823, 8)
(10824, 8)
(10825, 8)
(10826, 8)
(10827, 8)
(10828, 8)
(10829, 8)
(10830, 8)
(10831, 8)
(10832, 8)
(10833, 8)
(10834, 8)
(10835, 8)
(10836, 8)
(10837, 8)
(10838, 8)
(10839, 8)
(10840, 8)
(10841, 8)
(10842, 8)
(10843, 8)
(10844, 8)
(10845, 8)
(10846, 8)
(10847, 8)
(10848, 8)
(10849, 8)
(10850, 8)
(10851, 8)
(10852, 8)
(10853, 8)
(10854, 8)
(10855, 8)
(10856, 8)
(10857, 8)
(10858, 8)
(10859, 8)
(10860, 8)
(10861, 8)
(10862, 8)
(10863, 8)
(10864, 8)
(10865, 8)
(10866, 8)
(10867, 8)
(10868, 8)
(10869, 8)
(10870, 8)
(10871, 8)
(10872, 8)
(10873, 8)
(10874, 8)
(10875, 8)
(10876, 8)
(10877, 8)
(10878, 8)
(10879, 8)
(10880, 8)
(10881, 8)
(10882, 8)
(10883, 8)
(10884, 8)
(10885, 8)
(10886, 8)
(10887, 8)
(10888, 8)
(10889, 8)
(10890, 8)
(10891, 8)
(10892, 8)
(10893, 8)
(10894, 8)
(10895, 8)
(10896, 8)
(10897, 8)
(10898, 8)
(10899, 8)
(10900, 8)
(10901, 8)
(10902, 8)
(10903, 8)
(10904, 8)
(10905, 8)

(11577, 8)
(11578, 8)
(11579, 8)
(11580, 8)
(11581, 8)
(11582, 8)
(11583, 8)
(11584, 8)
(11585, 8)
(11586, 8)
(11587, 8)
(11588, 8)
(11589, 8)
(11590, 8)
(11591, 8)
(11592, 8)
(11593, 8)
(11594, 8)
(11595, 8)
(11596, 8)
(11597, 8)
(11598, 8)
(11599, 8)
(11600, 8)
(11601, 8)
(11602, 8)
(11603, 8)
(11604, 8)
(11605, 8)
(11606, 8)
(11607, 8)
(11608, 8)
(11609, 8)
(11610, 8)
(11611, 8)
(11612, 8)
(11613, 8)
(11614, 8)
(11615, 8)
(11616, 8)
(11617, 8)
(11618, 8)
(11619, 8)
(11620, 8)
(11621, 8)
(11622, 8)
(11623, 8)
(11624, 8)
(11625, 8)
(11626, 8)
(11627, 8)
(11628, 8)
(11629, 8)
(11630, 8)
(11631, 8)
(11632, 8)
(11633, 8)
(11634, 8)
(11635, 8)
(11636, 8)
(11637, 8)
(11638, 8)
(11639, 8)
(11640, 8)
(11641, 8)
(11642, 8)
(11643, 8)
(11644, 8)
(11645, 8)
(11646, 8)
(11647, 8)
(11648, 8)
(11649, 8)
(11650, 8)
(11651, 8)
(11652, 8)
(11653, 8)
(11654, 8)
(11655, 8)
(11656, 8)
(11657, 8)
(11658, 8)
(11659, 8)
(11660, 8)
(11661, 8)
(11662, 8)
(11663, 8)
(11664, 8)
(11665, 8)
(11666, 8)
(11667, 8)

(12344, 8)
(12345, 8)
(12346, 8)
(12347, 8)
(12348, 8)
(12349, 8)
(12350, 8)
(12351, 8)
(12352, 8)
(12353, 8)
(12354, 8)
(12355, 8)
(12356, 8)
(12357, 8)
(12358, 8)
(12359, 8)
(12360, 8)
(12361, 8)
(12362, 8)
(12363, 8)
(12364, 8)
(12365, 8)
(12366, 8)
(12367, 8)
(12368, 8)
(12369, 8)
(12370, 8)
(12371, 8)
(12372, 8)
(12373, 8)
(12374, 8)
(12375, 8)
(12376, 8)
(12377, 8)
(12378, 8)
(12379, 8)
(12380, 8)
(12381, 8)
(12382, 8)
(12383, 8)
(12384, 8)
(12385, 8)
(12386, 8)
(12387, 8)
(12388, 8)
(12389, 8)
(12390, 8)
(12391, 8)
(12392, 8)
(12393, 8)
(12394, 8)
(12395, 8)
(12396, 8)
(12397, 8)
(12398, 8)
(12399, 8)
(12400, 8)
(12401, 8)
(12402, 8)
(12403, 8)
(12404, 8)
(12405, 8)
(12406, 8)
(12407, 8)
(12408, 8)
(12409, 8)
(12410, 8)
(12411, 8)
(12412, 8)
(12413, 8)
(12414, 8)
(12415, 8)
(12416, 8)
(12417, 8)
(12418, 8)
(12419, 8)
(12420, 8)
(12421, 8)
(12422, 8)
(12423, 8)
(12424, 8)
(12425, 8)
(12426, 8)
(12427, 8)
(12428, 8)
(12429, 8)
(12430, 8)
(12431, 8)
(12432, 8)
(12433, 8)
(12434, 8)

(13104, 8)
(13105, 8)
(13106, 8)
(13107, 8)
(13108, 8)
(13109, 8)
(13110, 8)
(13111, 8)
(13112, 8)
(13113, 8)
(13114, 8)
(13115, 8)
(13116, 8)
(13117, 8)
(13118, 8)
(13119, 8)
(13120, 8)
(13121, 8)
(13122, 8)
(13123, 8)
(13124, 8)
(13125, 8)
(13126, 8)
(13127, 8)
(13128, 8)
(13129, 8)
(13130, 8)
(13131, 8)
(13132, 8)
(13133, 8)
(13134, 8)
(13135, 8)
(13136, 8)
(13137, 8)
(13138, 8)
(13139, 8)
(13140, 8)
(13141, 8)
(13142, 8)
(13143, 8)
(13144, 8)
(13145, 8)
(13146, 8)
(13147, 8)
(13148, 8)
(13149, 8)
(13150, 8)
(13151, 8)
(13152, 8)
(13153, 8)
(13154, 8)
(13155, 8)
(13156, 8)
(13157, 8)
(13158, 8)
(13159, 8)
(13160, 8)
(13161, 8)
(13162, 8)
(13163, 8)
(13164, 8)
(13165, 8)
(13166, 8)
(13167, 8)
(13168, 8)
(13169, 8)
(13170, 8)
(13171, 8)
(13172, 8)
(13173, 8)
(13174, 8)
(13175, 8)
(13176, 8)
(13177, 8)
(13178, 8)
(13179, 8)
(13180, 8)
(13181, 8)
(13182, 8)
(13183, 8)
(13184, 8)
(13185, 8)
(13186, 8)
(13187, 8)
(13188, 8)
(13189, 8)
(13190, 8)
(13191, 8)
(13192, 8)
(13193, 8)
(13194, 8)

(13859, 8)
(13860, 8)
(13861, 8)
(13862, 8)
(13863, 8)
(13864, 8)
(13865, 8)
(13866, 8)
(13867, 8)
(13868, 8)
(13869, 8)
(13870, 8)
(13871, 8)
(13872, 8)
(13873, 8)
(13874, 8)
(13875, 8)
(13876, 8)
(13877, 8)
(13878, 8)
(13879, 8)
(13880, 8)
(13881, 8)
(13882, 8)
(13883, 8)
(13884, 8)
(13885, 8)
(13886, 8)
(13887, 8)
(13888, 8)
(13889, 8)
(13890, 8)
(13891, 8)
(13892, 8)
(13893, 8)
(13894, 8)
(13895, 8)
(13896, 8)
(13897, 8)
(13898, 8)
(13899, 8)
(13900, 8)
(13901, 8)
(13902, 8)
(13903, 8)
(13904, 8)
(13905, 8)
(13906, 8)
(13907, 8)
(13908, 8)
(13909, 8)
(13910, 8)
(13911, 8)
(13912, 8)
(13913, 8)
(13914, 8)
(13915, 8)
(13916, 8)
(13917, 8)
(13918, 8)
(13919, 8)
(13920, 8)
(13921, 8)
(13922, 8)
(13923, 8)
(13924, 8)
(13925, 8)
(13926, 8)
(13927, 8)
(13928, 8)
(13929, 8)
(13930, 8)
(13931, 8)
(13932, 8)
(13933, 8)
(13934, 8)
(13935, 8)
(13936, 8)
(13937, 8)
(13938, 8)
(13939, 8)
(13940, 8)
(13941, 8)
(13942, 8)
(13943, 8)
(13944, 8)
(13945, 8)
(13946, 8)
(13947, 8)
(13948, 8)
(13949, 8)

(14604, 8)
(14605, 8)
(14606, 8)
(14607, 8)
(14608, 8)
(14609, 8)
(14610, 8)
(14611, 8)
(14612, 8)
(14613, 8)
(14614, 8)
(14615, 8)
(14616, 8)
(14617, 8)
(14618, 8)
(14619, 8)
(14620, 8)
(14621, 8)
(14622, 8)
(14623, 8)
(14624, 8)
(14625, 8)
(14626, 8)
(14627, 8)
(14628, 8)
(14629, 8)
(14630, 8)
(14631, 8)
(14632, 8)
(14633, 8)
(14634, 8)
(14635, 8)
(14636, 8)
(14637, 8)
(14638, 8)
(14639, 8)
(14640, 8)
(14641, 8)
(14642, 8)
(14643, 8)
(14644, 8)
(14645, 8)
(14646, 8)
(14647, 8)
(14648, 8)
(14649, 8)
(14650, 8)
(14651, 8)
(14652, 8)
(14653, 8)
(14654, 8)
(14655, 8)
(14656, 8)
(14657, 8)
(14658, 8)
(14659, 8)
(14660, 8)
(14661, 8)
(14662, 8)
(14663, 8)
(14664, 8)
(14665, 8)
(14666, 8)
(14667, 8)
(14668, 8)
(14669, 8)
(14670, 8)
(14671, 8)
(14672, 8)
(14673, 8)
(14674, 8)
(14675, 8)
(14676, 8)
(14677, 8)
(14678, 8)
(14679, 8)
(14680, 8)
(14681, 8)
(14682, 8)
(14683, 8)
(14684, 8)
(14685, 8)
(14686, 8)
(14687, 8)
(14688, 8)
(14689, 8)
(14690, 8)
(14691, 8)
(14692, 8)
(14693, 8)
(14694, 8)

(15363, 8)
(15364, 8)
(15365, 8)
(15366, 8)
(15367, 8)
(15368, 8)
(15369, 8)
(15370, 8)
(15371, 8)
(15372, 8)
(15373, 8)
(15374, 8)
(15375, 8)
(15376, 8)
(15377, 8)
(15378, 8)
(15379, 8)
(15380, 8)
(15381, 8)
(15382, 8)
(15383, 8)
(15384, 8)
(15385, 8)
(15386, 8)
(15387, 8)
(15388, 8)
(15389, 8)
(15390, 8)
(15391, 8)
(15392, 8)
(15393, 8)
(15394, 8)
(15395, 8)
(15396, 8)
(15397, 8)
(15398, 8)
(15399, 8)
(15400, 8)
(15401, 8)
(15402, 8)
(15403, 8)
(15404, 8)
(15405, 8)
(15406, 8)
(15407, 8)
(15408, 8)
(15409, 8)
(15410, 8)
(15411, 8)
(15412, 8)
(15413, 8)
(15414, 8)
(15415, 8)
(15416, 8)
(15417, 8)
(15418, 8)
(15419, 8)
(15420, 8)
(15421, 8)
(15422, 8)
(15423, 8)
(15424, 8)
(15425, 8)
(15426, 8)
(15427, 8)
(15428, 8)
(15429, 8)
(15430, 8)
(15431, 8)
(15432, 8)
(15433, 8)
(15434, 8)
(15435, 8)
(15436, 8)
(15437, 8)
(15438, 8)
(15439, 8)
(15440, 8)
(15441, 8)
(15442, 8)
(15443, 8)
(15444, 8)
(15445, 8)
(15446, 8)
(15447, 8)
(15448, 8)
(15449, 8)
(15450, 8)
(15451, 8)
(15452, 8)
(15453, 8)

(16122, 8)
(16123, 8)
(16124, 8)
(16125, 8)
(16126, 8)
(16127, 8)
(16128, 8)
(16129, 8)
(16130, 8)
(16131, 8)
(16132, 8)
(16133, 8)
(16134, 8)
(16135, 8)
(16136, 8)
(16137, 8)
(16138, 8)
(16139, 8)
(16140, 8)
(16141, 8)
(16142, 8)
(16143, 8)
(16144, 8)
(16145, 8)
(16146, 8)
(16147, 8)
(16148, 8)
(16149, 8)
(16150, 8)
(16151, 8)
(16152, 8)
(16153, 8)
(16154, 8)
(16155, 8)
(16156, 8)
(16157, 8)
(16158, 8)
(16159, 8)
(16160, 8)
(16161, 8)
(16162, 8)
(16163, 8)
(16164, 8)
(16165, 8)
(16166, 8)
(16167, 8)
(16168, 8)
(16169, 8)
(16170, 8)
(16171, 8)
(16172, 8)
(16173, 8)
(16174, 8)
(16175, 8)
(16176, 8)
(16177, 8)
(16178, 8)
(16179, 8)
(16180, 8)
(16181, 8)
(16182, 8)
(16183, 8)
(16184, 8)
(16185, 8)
(16186, 8)
(16187, 8)
(16188, 8)
(16189, 8)
(16190, 8)
(16191, 8)
(16192, 8)
(16193, 8)
(16194, 8)
(16195, 8)
(16196, 8)
(16197, 8)
(16198, 8)
(16199, 8)
(16200, 8)
(16201, 8)
(16202, 8)
(16203, 8)
(16204, 8)
(16205, 8)
(16206, 8)
(16207, 8)
(16208, 8)
(16209, 8)
(16210, 8)
(16211, 8)
(16212, 8)

(16881, 8)
(16882, 8)
(16883, 8)
(16884, 8)
(16885, 8)
(16886, 8)
(16887, 8)
(16888, 8)
(16889, 8)
(16890, 8)
(16891, 8)
(16892, 8)
(16893, 8)
(16894, 8)
(16895, 8)
(16896, 8)
(16897, 8)
(16898, 8)
(16899, 8)
(16900, 8)
(16901, 8)
(16902, 8)
(16903, 8)
(16904, 8)
(16905, 8)
(16906, 8)
(16907, 8)
(16908, 8)
(16909, 8)
(16910, 8)
(16911, 8)
(16912, 8)
(16913, 8)
(16914, 8)
(16915, 8)
(16916, 8)
(16917, 8)
(16918, 8)
(16919, 8)
(16920, 8)
(16921, 8)
(16922, 8)
(16923, 8)
(16924, 8)
(16925, 8)
(16926, 8)
(16927, 8)
(16928, 8)
(16929, 8)
(16930, 8)
(16931, 8)
(16932, 8)
(16933, 8)
(16934, 8)
(16935, 8)
(16936, 8)
(16937, 8)
(16938, 8)
(16939, 8)
(16940, 8)
(16941, 8)
(16942, 8)
(16943, 8)
(16944, 8)
(16945, 8)
(16946, 8)
(16947, 8)
(16948, 8)
(16949, 8)
(16950, 8)
(16951, 8)
(16952, 8)
(16953, 8)
(16954, 8)
(16955, 8)
(16956, 8)
(16957, 8)
(16958, 8)
(16959, 8)
(16960, 8)
(16961, 8)
(16962, 8)
(16963, 8)
(16964, 8)
(16965, 8)
(16966, 8)
(16967, 8)
(16968, 8)
(16969, 8)
(16970, 8)
(16971, 8)

(17643, 8)
(17644, 8)
(17645, 8)
(17646, 8)
(17647, 8)
(17648, 8)
(17649, 8)
(17650, 8)
(17651, 8)
(17652, 8)
(17653, 8)
(17654, 8)
(17655, 8)
(17656, 8)
(17657, 8)
(17658, 8)
(17659, 8)
(17660, 8)
(17661, 8)
(17662, 8)
(17663, 8)
(17664, 8)
(17665, 8)
(17666, 8)
(17667, 8)
(17668, 8)
(17669, 8)
(17670, 8)
(17671, 8)
(17672, 8)
(17673, 8)
(17674, 8)
(17675, 8)
(17676, 8)
(17677, 8)
(17678, 8)
(17679, 8)
(17680, 8)
(17681, 8)
(17682, 8)
(17683, 8)
(17684, 8)
(17685, 8)
(17686, 8)
(17687, 8)
(17688, 8)
(17689, 8)
(17690, 8)
(17691, 8)
(17692, 8)
(17693, 8)
(17694, 8)
(17695, 8)
(17696, 8)
(17697, 8)
(17698, 8)
(17699, 8)
(17700, 8)
(17701, 8)
(17702, 8)
(17703, 8)
(17704, 8)
(17705, 8)
(17706, 8)
(17707, 8)
(17708, 8)
(17709, 8)
(17710, 8)
(17711, 8)
(17712, 8)
(17713, 8)
(17714, 8)
(17715, 8)
(17716, 8)
(17717, 8)
(17718, 8)
(17719, 8)
(17720, 8)
(17721, 8)
(17722, 8)
(17723, 8)
(17724, 8)
(17725, 8)
(17726, 8)
(17727, 8)
(17728, 8)
(17729, 8)
(17730, 8)
(17731, 8)
(17732, 8)
(17733, 8)

(18410, 8)
(18411, 8)
(18412, 8)
(18413, 8)
(18414, 8)
(18415, 8)
(18416, 8)
(18417, 8)
(18418, 8)
(18419, 8)
(18420, 8)
(18421, 8)
(18422, 8)
(18423, 8)
(18424, 8)
(18425, 8)
(18426, 8)
(18427, 8)
(18428, 8)
(18429, 8)
(18430, 8)
(18431, 8)
(18432, 8)
(18433, 8)
(18434, 8)
(18435, 8)
(18436, 8)
(18437, 8)
(18438, 8)
(18439, 8)
(18440, 8)
(18441, 8)
(18442, 8)
(18443, 8)
(18444, 8)
(18445, 8)
(18446, 8)
(18447, 8)
(18448, 8)
(18449, 8)
(18450, 8)
(18451, 8)
(18452, 8)
(18453, 8)
(18454, 8)
(18455, 8)
(18456, 8)
(18457, 8)
(18458, 8)
(18459, 8)
(18460, 8)
(18461, 8)
(18462, 8)
(18463, 8)
(18464, 8)
(18465, 8)
(18466, 8)
(18467, 8)
(18468, 8)
(18469, 8)
(18470, 8)
(18471, 8)
(18472, 8)
(18473, 8)
(18474, 8)
(18475, 8)
(18476, 8)
(18477, 8)
(18478, 8)
(18479, 8)
(18480, 8)
(18481, 8)
(18482, 8)
(18483, 8)
(18484, 8)
(18485, 8)
(18486, 8)
(18487, 8)
(18488, 8)
(18489, 8)
(18490, 8)
(18491, 8)
(18492, 8)
(18493, 8)
(18494, 8)
(18495, 8)
(18496, 8)
(18497, 8)
(18498, 8)
(18499, 8)
(18500, 8)

(19160, 8)
(19161, 8)
(19162, 8)
(19163, 8)
(19164, 8)
(19165, 8)
(19166, 8)
(19167, 8)
(19168, 8)
(19169, 8)
(19170, 8)
(19171, 8)
(19172, 8)
(19173, 8)
(19174, 8)
(19175, 8)
(19176, 8)
(19177, 8)
(19178, 8)
(19179, 8)
(19180, 8)
(19181, 8)
(19182, 8)
(19183, 8)
(19184, 8)
(19185, 8)
(19186, 8)
(19187, 8)
(19188, 8)
(19189, 8)
(19190, 8)
(19191, 8)
(19192, 8)
(19193, 8)
(19194, 8)
(19195, 8)
(19196, 8)
(19197, 8)
(19198, 8)
(19199, 8)
(19200, 8)
(19201, 8)
(19202, 8)
(19203, 8)
(19204, 8)
(19205, 8)
(19206, 8)
(19207, 8)
(19208, 8)
(19209, 8)
(19210, 8)
(19211, 8)
(19212, 8)
(19213, 8)
(19214, 8)
(19215, 8)
(19216, 8)
(19217, 8)
(19218, 8)
(19219, 8)
(19220, 8)
(19221, 8)
(19222, 8)
(19223, 8)
(19224, 8)
(19225, 8)
(19226, 8)
(19227, 8)
(19228, 8)
(19229, 8)
(19230, 8)
(19231, 8)
(19232, 8)
(19233, 8)
(19234, 8)
(19235, 8)
(19236, 8)
(19237, 8)
(19238, 8)
(19239, 8)
(19240, 8)
(19241, 8)
(19242, 8)
(19243, 8)
(19244, 8)
(19245, 8)
(19246, 8)
(19247, 8)
(19248, 8)
(19249, 8)
(19250, 8)

(19922, 8)
(19923, 8)
(19924, 8)
(19925, 8)
(19926, 8)
(19927, 8)
(19928, 8)
(19929, 8)
(19930, 8)
(19931, 8)
(19932, 8)
(19933, 8)
(19934, 8)
(19935, 8)
(19936, 8)
(19937, 8)
(19938, 8)
(19939, 8)
(19940, 8)
(19941, 8)
(19942, 8)
(19943, 8)
(19944, 8)
(19945, 8)
(19946, 8)
(19947, 8)
(19948, 8)
(19949, 8)
(19950, 8)
(19951, 8)
(19952, 8)
(19953, 8)
(19954, 8)
(19955, 8)
(19956, 8)
(19957, 8)
(19958, 8)
(19959, 8)
(19960, 8)
(19961, 8)
(19962, 8)
(19963, 8)
(19964, 8)
(19965, 8)
(19966, 8)
(19967, 8)
(19968, 8)
(19969, 8)
(19970, 8)
(19971, 8)
(19972, 8)
(19973, 8)
(19974, 8)
(19975, 8)
(19976, 8)
(19977, 8)
(19978, 8)
(19979, 8)
(19980, 8)
(19981, 8)
(19982, 8)
(19983, 8)
(19984, 8)
(19985, 8)
(19986, 8)
(19987, 8)
(19988, 8)
(19989, 8)
(19990, 8)
(19991, 8)
(19992, 8)
(19993, 8)
(19994, 8)
(19995, 8)
(19996, 8)
(19997, 8)
(19998, 8)
(19999, 8)
(20000, 8)
(20001, 8)
(20002, 8)
(20003, 8)
(20004, 8)
(20005, 8)
(20006, 8)
(20007, 8)
(20008, 8)
(20009, 8)
(20010, 8)
(20011, 8)
(20012, 8)

(20669, 8)
(20670, 8)
(20671, 8)
(20672, 8)
(20673, 8)
(20674, 8)
(20675, 8)
(20676, 8)
(20677, 8)
(20678, 8)
(20679, 8)
(20680, 8)
(20681, 8)
(20682, 8)
(20683, 8)
(20684, 8)
(20685, 8)
(20686, 8)
(20687, 8)
(20688, 8)
(20689, 8)
(20690, 8)
(20691, 8)
(20692, 8)
(20693, 8)
(20694, 8)
(20695, 8)
(20696, 8)
(20697, 8)
(20698, 8)
(20699, 8)
(20700, 8)
(20701, 8)
(20702, 8)
(20703, 8)
(20704, 8)
(20705, 8)
(20706, 8)
(20707, 8)
(20708, 8)
(20709, 8)
(20710, 8)
(20711, 8)
(20712, 8)
(20713, 8)
(20714, 8)
(20715, 8)
(20716, 8)
(20717, 8)
(20718, 8)
(20719, 8)
(20720, 8)
(20721, 8)
(20722, 8)
(20723, 8)
(20724, 8)
(20725, 8)
(20726, 8)
(20727, 8)
(20728, 8)
(20729, 8)
(20730, 8)
(20731, 8)
(20732, 8)
(20733, 8)
(20734, 8)
(20735, 8)
(20736, 8)
(20737, 8)
(20738, 8)
(20739, 8)
(20740, 8)
(20741, 8)
(20742, 8)
(20743, 8)
(20744, 8)
(20745, 8)
(20746, 8)
(20747, 8)
(20748, 8)
(20749, 8)
(20750, 8)
(20751, 8)
(20752, 8)
(20753, 8)
(20754, 8)
(20755, 8)
(20756, 8)
(20757, 8)
(20758, 8)
(20759, 8)

(21416, 8)
(21417, 8)
(21418, 8)
(21419, 8)
(21420, 8)
(21421, 8)
(21422, 8)
(21423, 8)
(21424, 8)
(21425, 8)
(21426, 8)
(21427, 8)
(21428, 8)
(21429, 8)
(21430, 8)
(21431, 8)
(21432, 8)
(21433, 8)
(21434, 8)
(21435, 8)
(21436, 8)
(21437, 8)
(21438, 8)
(21439, 8)
(21440, 8)
(21441, 8)
(21442, 8)
(21443, 8)
(21444, 8)
(21445, 8)
(21446, 8)
(21447, 8)
(21448, 8)
(21449, 8)
(21450, 8)
(21451, 8)
(21452, 8)
(21453, 8)
(21454, 8)
(21455, 8)
(21456, 8)
(21457, 8)
(21458, 8)
(21459, 8)
(21460, 8)
(21461, 8)
(21462, 8)
(21463, 8)
(21464, 8)
(21465, 8)
(21466, 8)
(21467, 8)
(21468, 8)
(21469, 8)
(21470, 8)
(21471, 8)
(21472, 8)
(21473, 8)
(21474, 8)
(21475, 8)
(21476, 8)
(21477, 8)
(21478, 8)
(21479, 8)
(21480, 8)
(21481, 8)
(21482, 8)
(21483, 8)
(21484, 8)
(21485, 8)
(21486, 8)
(21487, 8)
(21488, 8)
(21489, 8)
(21490, 8)
(21491, 8)
(21492, 8)
(21493, 8)
(21494, 8)
(21495, 8)
(21496, 8)
(21497, 8)
(21498, 8)
(21499, 8)
(21500, 8)
(21501, 8)
(21502, 8)
(21503, 8)
(21504, 8)
(21505, 8)
(21506, 8)

(22167, 8)
(22168, 8)
(22169, 8)
(22170, 8)
(22171, 8)
(22172, 8)
(22173, 8)
(22174, 8)
(22175, 8)
(22176, 8)
(22177, 8)
(22178, 8)
(22179, 8)
(22180, 8)
(22181, 8)
(22182, 8)
(22183, 8)
(22184, 8)
(22185, 8)
(22186, 8)
(22187, 8)
(22188, 8)
(22189, 8)
(22190, 8)
(22191, 8)
(22192, 8)
(22193, 8)
(22194, 8)
(22195, 8)
(22196, 8)
(22197, 8)
(22198, 8)
(22199, 8)
(22200, 8)
(22201, 8)
(22202, 8)
(22203, 8)
(22204, 8)
(22205, 8)
(22206, 8)
(22207, 8)
(22208, 8)
(22209, 8)
(22210, 8)
(22211, 8)
(22212, 8)
(22213, 8)
(22214, 8)
(22215, 8)
(22216, 8)
(22217, 8)
(22218, 8)
(22219, 8)
(22220, 8)
(22221, 8)
(22222, 8)
(22223, 8)
(22224, 8)
(22225, 8)
(22226, 8)
(22227, 8)
(22228, 8)
(22229, 8)
(22230, 8)
(22231, 8)
(22232, 8)
(22233, 8)
(22234, 8)
(22235, 8)
(22236, 8)
(22237, 8)
(22238, 8)
(22239, 8)
(22240, 8)
(22241, 8)
(22242, 8)
(22243, 8)
(22244, 8)
(22245, 8)
(22246, 8)
(22247, 8)
(22248, 8)
(22249, 8)
(22250, 8)
(22251, 8)
(22252, 8)
(22253, 8)
(22254, 8)
(22255, 8)
(22256, 8)
(22257, 8)

(22912, 8)
(22913, 8)
(22914, 8)
(22915, 8)
(22916, 8)
(22917, 8)
(22918, 8)
(22919, 8)
(22920, 8)
(22921, 8)
(22922, 8)
(22923, 8)
(22924, 8)
(22925, 8)
(22926, 8)
(22927, 8)
(22928, 8)
(22929, 8)
(22930, 8)
(22931, 8)
(22932, 8)
(22933, 8)
(22934, 8)
(22935, 8)
(22936, 8)
(22937, 8)
(22938, 8)
(22939, 8)
(22940, 8)
(22941, 8)
(22942, 8)
(22943, 8)
(22944, 8)
(22945, 8)
(22946, 8)
(22947, 8)
(22948, 8)
(22949, 8)
(22950, 8)
(22951, 8)
(22952, 8)
(22953, 8)
(22954, 8)
(22955, 8)
(22956, 8)
(22957, 8)
(22958, 8)
(22959, 8)
(22960, 8)
(22961, 8)
(22962, 8)
(22963, 8)
(22964, 8)
(22965, 8)
(22966, 8)
(22967, 8)
(22968, 8)
(22969, 8)
(22970, 8)
(22971, 8)
(22972, 8)
(22973, 8)
(22974, 8)
(22975, 8)
(22976, 8)
(22977, 8)
(22978, 8)
(22979, 8)
(22980, 8)
(22981, 8)
(22982, 8)
(22983, 8)
(22984, 8)
(22985, 8)
(22986, 8)
(22987, 8)
(22988, 8)
(22989, 8)
(22990, 8)
(22991, 8)
(22992, 8)
(22993, 8)
(22994, 8)
(22995, 8)
(22996, 8)
(22997, 8)
(22998, 8)
(22999, 8)
(23000, 8)
(23001, 8)
(23002, 8)

(23663, 8)
(23664, 8)
(23665, 8)
(23666, 8)
(23667, 8)
(23668, 8)
(23669, 8)
(23670, 8)
(23671, 8)
(23672, 8)
(23673, 8)
(23674, 8)
(23675, 8)
(23676, 8)
(23677, 8)
(23678, 8)
(23679, 8)
(23680, 8)
(23681, 8)
(23682, 8)
(23683, 8)
(23684, 8)
(23685, 8)
(23686, 8)
(23687, 8)
(23688, 8)
(23689, 8)
(23690, 8)
(23691, 8)
(23692, 8)
(23693, 8)
(23694, 8)
(23695, 8)
(23696, 8)
(23697, 8)
(23698, 8)
(23699, 8)
(23700, 8)
(23701, 8)
(23702, 8)
(23703, 8)
(23704, 8)
(23705, 8)
(23706, 8)
(23707, 8)
(23708, 8)
(23709, 8)
(23710, 8)
(23711, 8)
(23712, 8)
(23713, 8)
(23714, 8)
(23715, 8)
(23716, 8)
(23717, 8)
(23718, 8)
(23719, 8)
(23720, 8)
(23721, 8)
(23722, 8)
(23723, 8)
(23724, 8)
(23725, 8)
(23726, 8)
(23727, 8)
(23728, 8)
(23729, 8)
(23730, 8)
(23731, 8)
(23732, 8)
(23733, 8)
(23734, 8)
(23735, 8)
(23736, 8)
(23737, 8)
(23738, 8)
(23739, 8)
(23740, 8)
(23741, 8)
(23742, 8)
(23743, 8)
(23744, 8)
(23745, 8)
(23746, 8)
(23747, 8)
(23748, 8)
(23749, 8)
(23750, 8)
(23751, 8)
(23752, 8)
(23753, 8)

(24426, 8)
(24427, 8)
(24428, 8)
(24429, 8)
(24430, 8)
(24431, 8)
(24432, 8)
(24433, 8)
(24434, 8)
(24435, 8)
(24436, 8)
(24437, 8)
(24438, 8)
(24439, 8)
(24440, 8)
(24441, 8)
(24442, 8)
(24443, 8)
(24444, 8)
(24445, 8)
(24446, 8)
(24447, 8)
(24448, 8)
(24449, 8)
(24450, 8)
(24451, 8)
(24452, 8)
(24453, 8)
(24454, 8)
(24455, 8)
(24456, 8)
(24457, 8)
(24458, 8)
(24459, 8)
(24460, 8)
(24461, 8)
(24462, 8)
(24463, 8)
(24464, 8)
(24465, 8)
(24466, 8)
(24467, 8)
(24468, 8)
(24469, 8)
(24470, 8)
(24471, 8)
(24472, 8)
(24473, 8)
(24474, 8)
(24475, 8)
(24476, 8)
(24477, 8)
(24478, 8)
(24479, 8)
(24480, 8)
(24481, 8)
(24482, 8)
(24483, 8)
(24484, 8)
(24485, 8)
(24486, 8)
(24487, 8)
(24488, 8)
(24489, 8)
(24490, 8)
(24491, 8)
(24492, 8)
(24493, 8)
(24494, 8)
(24495, 8)
(24496, 8)
(24497, 8)
(24498, 8)
(24499, 8)
(24500, 8)
(24501, 8)
(24502, 8)
(24503, 8)
(24504, 8)
(24505, 8)
(24506, 8)
(24507, 8)
(24508, 8)
(24509, 8)
(24510, 8)
(24511, 8)
(24512, 8)
(24513, 8)
(24514, 8)
(24515, 8)
(24516, 8)

(25188, 8)
(25189, 8)
(25190, 8)
(25191, 8)
(25192, 8)
(25193, 8)
(25194, 8)
(25195, 8)
(25196, 8)
(25197, 8)
(25198, 8)
(25199, 8)
(25200, 8)
(25201, 8)
(25202, 8)
(25203, 8)
(25204, 8)
(25205, 8)
(25206, 8)
(25207, 8)
(25208, 8)
(25209, 8)
(25210, 8)
(25211, 8)
(25212, 8)
(25213, 8)
(25214, 8)
(25215, 8)
(25216, 8)
(25217, 8)
(25218, 8)
(25219, 8)
(25220, 8)
(25221, 8)
(25222, 8)
(25223, 8)
(25224, 8)
(25225, 8)
(25226, 8)
(25227, 8)
(25228, 8)
(25229, 8)
(25230, 8)
(25231, 8)
(25232, 8)
(25233, 8)
(25234, 8)
(25235, 8)
(25236, 8)
(25237, 8)
(25238, 8)
(25239, 8)
(25240, 8)
(25241, 8)
(25242, 8)
(25243, 8)
(25244, 8)
(25245, 8)
(25246, 8)
(25247, 8)
(25248, 8)
(25249, 8)
(25250, 8)
(25251, 8)
(25252, 8)
(25253, 8)
(25254, 8)
(25255, 8)
(25256, 8)
(25257, 8)
(25258, 8)
(25259, 8)
(25260, 8)
(25261, 8)
(25262, 8)
(25263, 8)
(25264, 8)
(25265, 8)
(25266, 8)
(25267, 8)
(25268, 8)
(25269, 8)
(25270, 8)
(25271, 8)
(25272, 8)
(25273, 8)
(25274, 8)
(25275, 8)
(25276, 8)
(25277, 8)
(25278, 8)

(25934, 8)
(25935, 8)
(25936, 8)
(25937, 8)
(25938, 8)
(25939, 8)
(25940, 8)
(25941, 8)
(25942, 8)
(25943, 8)
(25944, 8)
(25945, 8)
(25946, 8)
(25947, 8)
(25948, 8)
(25949, 8)
(25950, 8)
(25951, 8)
(25952, 8)
(25953, 8)
(25954, 8)
(25955, 8)
(25956, 8)
(25957, 8)
(25958, 8)
(25959, 8)
(25960, 8)
(25961, 8)
(25962, 8)
(25963, 8)
(25964, 8)
(25965, 8)
(25966, 8)
(25967, 8)
(25968, 8)
(25969, 8)
(25970, 8)
(25971, 8)
(25972, 8)
(25973, 8)
(25974, 8)
(25975, 8)
(25976, 8)
(25977, 8)
(25978, 8)
(25979, 8)
(25980, 8)
(25981, 8)
(25982, 8)
(25983, 8)
(25984, 8)
(25985, 8)
(25986, 8)
(25987, 8)
(25988, 8)
(25989, 8)
(25990, 8)
(25991, 8)
(25992, 8)
(25993, 8)
(25994, 8)
(25995, 8)
(25996, 8)
(25997, 8)
(25998, 8)
(25999, 8)
(26000, 8)
(26001, 8)
(26002, 8)
(26003, 8)
(26004, 8)
(26005, 8)
(26006, 8)
(26007, 8)
(26008, 8)
(26009, 8)
(26010, 8)
(26011, 8)
(26012, 8)
(26013, 8)
(26014, 8)
(26015, 8)
(26016, 8)
(26017, 8)
(26018, 8)
(26019, 8)
(26020, 8)
(26021, 8)
(26022, 8)
(26023, 8)
(26024, 8)

(26682, 8)
(26683, 8)
(26684, 8)
(26685, 8)
(26686, 8)
(26687, 8)
(26688, 8)
(26689, 8)
(26690, 8)
(26691, 8)
(26692, 8)
(26693, 8)
(26694, 8)
(26695, 8)
(26696, 8)
(26697, 8)
(26698, 8)
(26699, 8)
(26700, 8)
(26701, 8)
(26702, 8)
(26703, 8)
(26704, 8)
(26705, 8)
(26706, 8)
(26707, 8)
(26708, 8)
(26709, 8)
(26710, 8)
(26711, 8)
(26712, 8)
(26713, 8)
(26714, 8)
(26715, 8)
(26716, 8)
(26717, 8)
(26718, 8)
(26719, 8)
(26720, 8)
(26721, 8)
(26722, 8)
(26723, 8)
(26724, 8)
(26725, 8)
(26726, 8)
(26727, 8)
(26728, 8)
(26729, 8)
(26730, 8)
(26731, 8)
(26732, 8)
(26733, 8)
(26734, 8)
(26735, 8)
(26736, 8)
(26737, 8)
(26738, 8)
(26739, 8)
(26740, 8)
(26741, 8)
(26742, 8)
(26743, 8)
(26744, 8)
(26745, 8)
(26746, 8)
(26747, 8)
(26748, 8)
(26749, 8)
(26750, 8)
(26751, 8)
(26752, 8)
(26753, 8)
(26754, 8)
(26755, 8)
(26756, 8)
(26757, 8)
(26758, 8)
(26759, 8)
(26760, 8)
(26761, 8)
(26762, 8)
(26763, 8)
(26764, 8)
(26765, 8)
(26766, 8)
(26767, 8)
(26768, 8)
(26769, 8)
(26770, 8)
(26771, 8)
(26772, 8)

In [58]:
print("Synthetic records needed: {}".format(n_synthetic_required_for(3))

17804

### 4-Anonymous Results

In [ ]:
synth_k4_df = synthetic_df(df, ['user_id', 'completed', 'explored'], quasi_list, 4)
print_status_stats(synth_k4_df)

In [ ]:
print("Synthetic records needed: {}".format(n_synthetic_required_for(4))

### 5-Anonymous Results

In [ ]:
synth_k5_df = synthetic_df(df, ['user_id', 'completed', 'explored'], quasi_list, 5)
print_status_stats(synth_k5_df)

In [ ]:
print("Synthetic records needed: {}".format(n_synthetic_required_for(5))

## Question 5

We've decided to do generalization on each of the quasi-identifiers first, and then suppress any additional records that need it. This isn't a particularly sophisticated technique, and it's certainly not the optimization that was performed in Angiuli and Waldo's "Statistical Tradeoffs ...".

Here are the generalization strategies we're using:
  * Drop `cc_by_ip` as it's largely redundant with `countryLabel`
  * Starting with the most specific, set any geographic field with fewer than 5000 records to "NA"
  * Change `YoB` to `birth_decade` which is the first three digits of the year
  * Blur all the `nforum` variables together by summing them into a `nforum_actions` variable
  * Change all the `nforum` varibles to booleans indicating whether they are greater than zero
  * Change `email_domain` to only the first three letters
  * Change `LoE` to `attended_college`; a boolean indicating if the student went to college
  
Changing the names of some of these attributes means we need a new set of quasi-identifiers to work from:

In [82]:
GENERALIZED_QUASI_IDS = [
    'postalCode',
    'city',
    'subdivision',
    'region',
    'countryLabel',
    'continent',
    'email_domain',
    'course_id',
    'mode',
    'gender',
    'nforum_activites',
    'nforum_posts_>0',
    'nforum_votes_>0',
    'nforum_endorsed_>0',
    'nforum_threads_>0',
    'nforum_comments_>0',
    'nforum_pinned_>0',
    'nforum_events_>0',
    'birth_decade',
    'attended_college'
]

### Useful Functions

Here we write all our functions for generalization, blurring, suppression, and presentation of results.

In [83]:
def generalize_geo(df, min_per_geo=5000):
    g = df.copy()
    GEO_HIERARCHY = [
        'postalCode',
        'city',
        'subdivision',
        'region',
        'countryLabel',
        'continent']
    g = g.drop('cc_by_ip', axis=1)
    for field in GEO_HIERARCHY:
        value_counts = g[field].value_counts()
        values_to_drop = set(value_counts[value_counts < min_per_geo].index.values)
        g.loc[g[field].isin(values_to_drop), field] = na_replacements[field]
    return g

In [84]:
def generalize_education(df):
    g = df.copy()
    # Associate is sometimes not included as college, but we are including it.
    college_levels = set(['p', 'p_se', 'p_oth', 'm', 'b', 'a'])
    g['attended_college'] = g['LoE'].isin(college_levels)
    return g.drop('LoE', axis=1)

In [85]:
def generalize_birth_year(df):
    g = df.copy()
    g['birth_decade'] = df['YoB'].apply(lambda y: str(y)[:3] + '*' if y>-1 else 'NA')
    return g.drop('YoB', axis=1)

In [ ]:
def blur_forum_actions(df):
    g = df.copy()
    g['forum_actions'] = g['nforum_posts'] + g['nforum_votes'] + g['nforum_endorsed'] + g['nforum_threads'] + g['nforum_comments'] + g['nforum_pinned'] + g['nforum_events']
    return g

In [86]:
def generalize_forum_actions(df):
    g = df.copy()
    fields = ['nforum_posts',
        'nforum_votes',
        'nforum_endorsed',
        'nforum_threads',
        'nforum_comments',
        'nforum_pinned',
        'nforum_events']
    for field in fields:
        g[field + '_>0'] = g[field] > 0
        g = g.drop(field, axis=1)
    return g

In [87]:
def generalize_email_domains(df):
    g = df.copy()
    g['email_domain'] = g['email_domain'].apply(lambda e: e[:3] + '*')
    return g

In [88]:
def generalize_df(df):
    return generalize_geo(
        generalize_birth_year(
        generalize_education(
        generalize_email_domains(blur_forum_actions(
        generalize_forum_actions(df))))))

In [111]:
def suppress_to_k_anonymize(df, quasi_ids, k):
    return df.groupby(quasi_ids).filter(lambda g: g.shape[0] >= k)

In [73]:
def status_count(df, status):
    return df.query(status + ' == True').shape[0]

def status_proportion(df, status):
    return status_count(df, status) / df.shape[0] * 100

def print_status_stats(df):
    print(f'Total students: {df.shape[0]}\n'
          f'Students explored: {status_count(df, "explored")}'
          f' - {status_proportion(df, "explored"):.2f}%\n'
          f'Students completed: {status_count(df, "completed")}'
          f' - {status_proportion(df, "completed"):.2f}%')

### Raw Dataset Stats

In [75]:
print_status_stats(df)

Total students: 1141735
Students explored: 122258 - 10.71%
Students completed: 28142 - 2.46%


### 3-Anonymous Results

In [ ]:
generalized = generalize_df(df)

In [112]:
k3_df = suppress_to_k_anonymize(generalized, GENERALIZED_QUASI_IDS, 3)

In [114]:
print_status_stats(k3_df)

Total students: 596580
Students explored: 41138 - 6.90%
Students completed: 4120 - 0.69%


### 4-Anonymous Results

In [115]:
k4_df = suppress_to_k_anonymize(generalized, GENERALIZED_QUASI_IDS, 4)

In [116]:
print_status_stats(k4_df)

Total students: 531984
Students explored: 34513 - 6.49%
Students completed: 3102 - 0.58%


### 5-Anonymous Results

In [118]:
k5_df = suppress_to_k_anonymize(generalized, GENERALIZED_QUASI_IDS, 5)

In [119]:
print_status_stats(k5_df)

Total students: 487536
Students explored: 30230 - 6.20%
Students completed: 2492 - 0.51%


### Observations

We can see that there's a steep dropoff in the number and proportion of students that completed and explored the class between the raw dataset and the 3-anonymous set. Specifically, in the raw data 2.46% of students completed, whereas in the 3-anonymous dataset, only 0.69% of students completed. As we increase the level of k-anonymity, there are fewer students overall and the proportion of students that completed and explored continues to drop off. This demonstrates that making a dataset k-anonymous may seriously impact the statistical validity of the dataset.

## Question 6

After reviewing all the field, we believe the following could be sensitive: `completed`, `grade`, `nproblem_check`, `nshow_answer`, `cert_status`. To check for the l-diversity, we're taking the minimum number of unique values per field for each quasi-identifier group.

In [120]:
def calc_l_diversity(df, field):
    return df.groupby(GENERALIZED_QUASI_IDS).apply(lambda g: g[field].nunique(dropna=False)).min()

In [122]:
sensitive_fields = ['completed', 'grade', 'nproblem_check', 'nshow_answer', 'cert_status']
k3_l_diversity = {field: calc_l_diversity(k3_df, field) for field in sensitive_fields}
k4_l_diversity = {field: calc_l_diversity(k4_df, field) for field in sensitive_fields}
k5_l_diversity = {field: calc_l_diversity(k5_df, field) for field in sensitive_fields}

### 3-Anonymous l-diversity Results

In [123]:
k3_l_diversity

{'completed': 1,
 'grade': 1,
 'nproblem_check': 1,
 'nshow_answer': 1,
 'cert_status': 1}

### 4-Anonymous l-diversity Results

In [124]:
k4_l_diversity

{'completed': 1,
 'grade': 1,
 'nproblem_check': 1,
 'nshow_answer': 1,
 'cert_status': 1}

### 5-Anonymous l-diversity Results

In [125]:
k5_l_diversity

{'completed': 1,
 'grade': 1,
 'nproblem_check': 1,
 'nshow_answer': 1,
 'cert_status': 1}

### Observations

From the above results, we can see that after all our work, none of our datasets rise above 1-diversity for any of the fields we consider sensitive. The improve this, we would need to either consider suppressing entire quasi-identifier columns or creating synthetic records to add diversity to each quasi-identifier group.